arduino-cli lib install PacketSerial: nil pointer dereference

## Environment

```
Rubens-MBP:arduino-cli ruben$ uname -a
Darwin Rubens-MBP.home 20.3.0 Darwin Kernel Version 20.3.0: Thu Jan 21 00:07:06 PST 2021; root:xnu-7195.81.3~1/RELEASE_X86_64 x86_64

Rubens-MBP:arduino ruben$ arduino-cli version
arduino-cli alpha Version: 0.15.1 Commit: c7403ed2 Date: 2021-02-04T10:11:39Z
```

## Description

Hi devs, I tried to install the `PacketSerial` library which failed 

```
Rubens-MBP:arduino ruben$ arduino-cli lib install PacketSerial
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x44be30c]

goroutine 1 [running]:
go.bug.st/relaxed-semver.(*Version).CompareTo(0xc000f2a630, 0x0, 0xc00171f5e9)
        /go/pkg/mod/go.bug.st/relaxed-semver@v0.0.0-20190922224835-391e10178d18/version.go:115 +0x3c
go.bug.st/relaxed-semver.(*Version).GreaterThan(...)
        /go/pkg/mod/go.bug.st/relaxed-semver@v0.0.0-20190922224835-391e10178d18/version.go:202
github.com/arduino/arduino-cli/arduino/libraries/librariesindex.(*Index).FindLibraryUpdate(0xc0001c2000, 0xc000581500, 0x0)
        /home/build/arduino/libraries/librariesindex/index.go:139 +0x88
github.com/arduino/arduino-cli/commands/lib.listLibraries(0xc00032c040, 0x4de0000, 0xc0014431d0, 0xc0037285a0, 0x0)
        /home/build/commands/lib/list.go:130 +0xe8
github.com/arduino/arduino-cli/commands/lib.LibraryResolveDependencies(0x4df1cc0, 0xc000038088, 0xc0014431d0, 0x1, 0xc0001c2dc8, 0x1)
        /home/build/commands/lib/resolve_deps.go:39 +0x145
github.com/arduino/arduino-cli/cli/lib.runInstallCommand(0xc001713340, 0xc0016bdaf0, 0x1, 0x1)
        /home/build/cli/lib/install.go:131 +0xed1
github.com/spf13/cobra.(*Command).execute(0xc001713340, 0xc0016bdad0, 0x1, 0x1, 0xc001713340, 0xc0016bdad0)
        /go/pkg/mod/github.com/spf13/cobra@v1.0.1-0.20200710201246-675ae5f5a98c/command.go:846 +0x29d
github.com/spf13/cobra.(*Command).ExecuteC(0xc0000bf8c0, 0x0, 0xc001699b00, 0x0)
        /go/pkg/mod/github.com/spf13/cobra@v1.0.1-0.20200710201246-675ae5f5a98c/command.go:950 +0x349
github.com/spf13/cobra.(*Command).Execute(...)
        /go/pkg/mod/github.com/spf13/cobra@v1.0.1-0.20200710201246-675ae5f5a98c/command.go:887
main.main()
        /home/build/main.go:31 +0x85
```

## Analysis

This happens due to a missing nil check  in `FindLibraryUpdate` for the following library:

```
&libraries.Library{Name:"UIPEthernet", Author:"ntruchsess", Maintainer:"Norbert Truchsess <norbert.truchsess@t-online.de>", Sentence:"Ethernet library for ENC28J60", Paragraph:"implements the same API as stock Ethernet-lib. Just replace the include of Ethernet.h with UIPEthernet.h", Website:"https://github.com/ntruchsess/arduino_uip", Category:"Uncategorized", Architectures:[]string{"avr"}, Types:[]string(nil), InstallDir:(*paths.Path)(0xc001797ad0), SourceDir:(*paths.Path)(0xc001797ad0), UtilityDir:(*paths.Path)(0xc001663930), Location:3, ContainerPlatform:(*cores.PlatformRelease)(nil), Layout:0x0, RealName:"UIPEthernet", DotALinkage:false, Precompiled:false, PrecompiledWithSources:false, LDflags:"", IsLegacy:false, Version:(*semver.Version)(nil), License:"Unspecified", Properties:(*properties.Map)(0xc0019f0750), Examples:paths.PathList{(*paths.Path)(0xc001663c10), (*paths.Path)(0xc001663c30), (*paths.Path)(0xc001663c60), (*paths.Path)(0xc001663c80), (*paths.Path)(0xc001663ca0), (*paths.Path)(0xc001663cc0)}, declaredHeaders:[]string(nil), sourceHeaders:[]string(nil), CompatibleWith:map[string]bool(nil)}
```

Ups I mistakenly used the patched version to reproduce this. Now with the unpatched version I can reproduce it.

```
INFO[0000] arduino-cli version 0.15.1                   
INFO[0000] Checking signature                            index=/Users/ruben/Library/Arduino15/package_index.json signatureFile=/Users/ruben/Library/Arduino15/package_index.json.sig trusted=true
INFO[0000] Checking if CLI is Bundled into the IDE      
INFO[0000] Loading hardware from: /Users/ruben/Library/Arduino15/packages 
INFO[0000] Loading package arduino from: /Users/ruben/Library/Arduino15/packages/arduino/hardware 
INFO[0000] Checking signature                            error="opening signature file: open /Users/ruben/Library/Arduino15/packages/arduino/hardware/avr/1.8.3/installed.json.sig: no such file or directory" index=/Users/ruben/Library/Arduino15/packages/arduino/hardware/avr/1.8.3/installed.json signatureFile=/Users/ruben/Library/Arduino15/packages/arduino/hardware/avr/1.8.3/installed.json.sig
INFO[0000] Loaded platform                               platform="arduino:avr@1.8.3"
INFO[0000] Checking existence of 'tools' path: /Users/ruben/Library/Arduino15/packages/arduino/tools 
INFO[0000] Loading tools from dir: /Users/ruben/Library/Arduino15/packages/arduino/tools 
INFO[0000] Loaded tool                                   tool="arduino:arduinoOTA@1.3.0"
INFO[0000] Loaded tool                                   tool="arduino:avr-gcc@7.3.0-atmel3.6.1-arduino7"
INFO[0000] Loaded tool                                   tool="arduino:avrdude@6.3.0-arduino17"
INFO[0000] Loading package builtin from: /Users/ruben/Library/Arduino15/packages/builtin 
INFO[0000] Checking existence of 'tools' path: /Users/ruben/Library/Arduino15/packages/builtin/tools 
INFO[0000] Loading tools from dir: /Users/ruben/Library/Arduino15/packages/builtin/tools 
INFO[0000] Loaded tool                                   tool="builtin:ctags@5.8-arduino11"
INFO[0000] Loaded tool                                   tool="builtin:serial-discovery@1.1.0"
INFO[0000] Loading hardware from: /Users/ruben/Documents/Arduino/hardware 
INFO[0000] Loading package MySensors from: /Users/ruben/Documents/Arduino/hardware/MySensors 
INFO[0000] Loaded platform                               platform="MySensors:avr@"
INFO[0000] Adding libraries dir                          dir=/Users/ruben/Documents/Arduino/libraries location=user
INFO[0000] Adding libraries dir                          dir=/Users/ruben/Library/Arduino15/packages/arduino/hardware/avr/1.8.3/libraries location=platform
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x44be30c]

goroutine 1 [running]:
go.bug.st/relaxed-semver.(*Version).CompareTo(0xc000dba7e0, 0x0, 0xc0005aa369)
	/go/pkg/mod/go.bug.st/relaxed-semver@v0.0.0-20190922224835-391e10178d18/version.go:115 +0x3c
go.bug.st/relaxed-semver.(*Version).GreaterThan(...)
	/go/pkg/mod/go.bug.st/relaxed-semver@v0.0.0-20190922224835-391e10178d18/version.go:202
github.com/arduino/arduino-cli/arduino/libraries/librariesindex.(*Index).FindLibraryUpdate(0xc000120018, 0xc000582f00, 0x0)
	/home/build/arduino/libraries/librariesindex/index.go:139 +0x88
github.com/arduino/arduino-cli/commands/lib.listLibraries(0xc0017e3c00, 0x4de0000, 0xc003ad2320, 0xc0036da690, 0x0)
	/home/build/commands/lib/list.go:130 +0xe8
github.com/arduino/arduino-cli/commands/lib.LibraryResolveDependencies(0x4df1cc0, 0xc000124008, 0xc003ad2320, 0x2, 0xc000010988, 0x1)
	/home/build/commands/lib/resolve_deps.go:39 +0x145
github.com/arduino/arduino-cli/cli/lib.runInstallCommand(0xc001675340, 0xc000135b80, 0x1, 0x2)
	/home/build/cli/lib/install.go:131 +0xed1
github.com/spf13/cobra.(*Command).execute(0xc001675340, 0xc000135b60, 0x2, 0x2, 0xc001675340, 0xc000135b60)
	/go/pkg/mod/github.com/spf13/cobra@v1.0.1-0.20200710201246-675ae5f5a98c/command.go:846 +0x29d
github.com/spf13/cobra.(*Command).ExecuteC(0xc0001398c0, 0x0, 0xc0015f97a0, 0x0)
	/go/pkg/mod/github.com/spf13/cobra@v1.0.1-0.20200710201246-675ae5f5a98c/command.go:950 +0x349
github.com/spf13/cobra.(*Command).Execute(...)
	/go/pkg/mod/github.com/spf13/cobra@v1.0.1-0.20200710201246-675ae5f5a98c/command.go:887
main.main()
	/home/build/main.go:31 +0x85
```

--- a/arduino/libraries/librariesindex/index.go
+++ b/arduino/libraries/librariesindex/index.go
@@ -16,6 +16,7 @@
 package librariesindex
 
 import (
+	"fmt"
 	"sort"
 
 	"github.com/arduino/arduino-cli/arduino/libraries"
@@ -135,6 +136,10 @@ func (idx *Index) FindLibraryUpdate(lib *libraries.Library) *Release {
 	indexLib := idx.FindIndexedLibrary(lib)
 	if indexLib == nil {
 		return nil
+	}
+	if lib.Version == nil {
+		fmt.Printf("[WARN] version for library loaded from %s is nil\n", lib.InstallDir)
+		return indexLib.Latest
 	}
 	if indexLib.Latest.Version.GreaterThan(lib.Version) {
 		return indexLib.Latest

--- a/arduino/libraries/librariesindex/index.go
+++ b/arduino/libraries/librariesindex/index.go
@@ -136,6 +136,8 @@ func (idx *Index) FindLibraryUpdate(lib *libraries.Library) *Release {
 	if indexLib == nil {
 		return nil
 	}
+	// library.Version is nil when when the version field in
+	// a library descriptor is malformed and could not be parsed.
 	if lib.Version == nil {
 		return indexLib.Latest
 	}

--- a/arduino/libraries/librariesindex/index.go
+++ b/arduino/libraries/librariesindex/index.go
@@ -136,12 +136,9 @@ func (idx *Index) FindLibraryUpdate(lib *libraries.Library) *Release {
 	if indexLib == nil {
 		return nil
 	}
-	// library.Version is nil when when the version field in
-	// a library descriptor is malformed and could not be parsed.
-	if lib.Version == nil {
-		return indexLib.Latest
-	}
-	if indexLib.Latest.Version.GreaterThan(lib.Version) {
+	// If a library.properties has an invalid version property, usually empty or malformed,
+	// the latest available version is returned
+	if lib.Version == nil || indexLib.Latest.Version.GreaterThan(lib.Version) {
 		return indexLib.Latest
 	}
 	return nil

--- a/arduino/libraries/librariesindex/index.go
+++ b/arduino/libraries/librariesindex/index.go
@@ -16,7 +16,6 @@
 package librariesindex
 
 import (
-	"fmt"
 	"sort"
 
 	"github.com/arduino/arduino-cli/arduino/libraries"
@@ -138,7 +137,6 @@ func (idx *Index) FindLibraryUpdate(lib *libraries.Library) *Release {
 		return nil
 	}
 	if lib.Version == nil {
-		fmt.Printf("[WARN] version for library loaded from %s is nil\n", lib.InstallDir)
 		return indexLib.Latest
 	}
 	if indexLib.Latest.Version.GreaterThan(lib.Version) {

--- a/arduino/libraries/librariesindex/index_test.go
+++ b/arduino/libraries/librariesindex/index_test.go
@@ -78,6 +78,10 @@ func TestIndexer(t *testing.T) {
 	require.NotNil(t, rtcUpdate)
 	require.Equal(t, "RTCZero@1.6.0", rtcUpdate.String())
 
+	rtcUpdateNoVersion := index.FindLibraryUpdate(&libraries.Library{Name: "RTCZero", Version: nil})
+	require.NotNil(t, rtcUpdateNoVersion)
+	require.Equal(t, "RTCZero@1.6.0", rtcUpdateNoVersion.String())
+
 	rtcNoUpdate := index.FindLibraryUpdate(&libraries.Library{Name: "RTCZero", Version: semver.MustParse("3.0.0")})
 	require.Nil(t, rtcNoUpdate)
 

--- a/arduino/libraries/librariesmanager/install.go
+++ b/arduino/libraries/librariesmanager/install.go
@@ -50,7 +50,7 @@ func (lm *LibrariesManager) InstallPrerequisiteCheck(indexLibrary *librariesinde
 			if installedLib.Location != libraries.User {
 				continue
 			}
-			if installedLib.Version.Equal(indexLibrary.Version) {
+			if installedLib.Version != nil && installedLib.Version.Equal(indexLibrary.Version) {
 				return installedLib.InstallDir, nil, ErrAlreadyInstalled
 			}
 			replaced = installedLib

--- a/test/test_lib.py
+++ b/test/test_lib.py
@@ -703,3 +703,65 @@ def test_lib_examples_with_case_mismatch(run_command, data_dir):
     # Verifies sketches with wrong casing are not returned
     assert str(examples_path / "NonBlocking" / "OnDemandNonBlocking") not in examples
     assert str(examples_path / "OnDemand" / "OnDemandWebPortal") not in examples
+
+
+def test_lib_list_using_library_with_invalid_version(run_command, data_dir):
+    assert run_command("update")
+
+    # Install a library
+    assert run_command("lib install WiFi101@0.16.1")
+
+    # Verifies library is correctly returned
+    res = run_command("lib list --format json")
+    assert res.ok
+    data = json.loads(res.stdout)
+    assert len(data) == 1
+    assert "0.16.1" == data[0]["library"]["version"]
+
+    # Changes the version of the currently installed library so that it's
+    # invalid
+    lib_path = Path(data_dir, "libraries", "WiFi101")
+    Path(lib_path, "library.properties").write_text("version=1.0001")
+
+    # Verifies version is now empty
+    res = run_command("lib list --format json")
+    assert res.ok
+    data = json.loads(res.stdout)
+    assert len(data) == 1
+    assert "version" not in data[0]["library"]
+
+
+def test_lib_upgrade_using_library_with_invalid_version(run_command, data_dir):
+    assert run_command("update")
+
+    # Install a library
+    assert run_command("lib install WiFi101@0.16.1")
+
+    # Verifies library is correctly returned
+    res = run_command("lib list --format json")
+    assert res.ok
+    data = json.loads(res.stdout)
+    assert len(data) == 1
+    assert "0.16.1" == data[0]["library"]["version"]
+
+    # Changes the version of the currently installed library so that it's
+    # invalid
+    lib_path = Path(data_dir, "libraries", "WiFi101")
+    Path(lib_path, "library.properties").write_text("version=1.0001")
+
+    # Verifies version is now empty
+    res = run_command("lib list --format json")
+    assert res.ok
+    data = json.loads(res.stdout)
+    assert len(data) == 1
+    assert "version" not in data[0]["library"]
+
+    # Upgrade library
+    assert run_command("lib upgrade WiFi101")
+
+    # Verifies library has been updated
+    res = run_command("lib list --format json")
+    assert res.ok
+    data = json.loads(res.stdout)
+    assert len(data) == 1
+    assert "" != data[0]["library"]["version"]

--- a/test/test_outdated.py
+++ b/test/test_outdated.py
@@ -13,6 +13,8 @@
 # software without disclosing the source code of your own applications. To purchase
 # a commercial license, send an email to license@arduino.cc.
 
+from pathlib import Path
+
 
 def test_outdated(run_command):
     # Updates index for cores and libraries
@@ -33,3 +35,26 @@ def test_outdated(run_command):
     lines = [l.strip() for l in result.stdout.splitlines()]
     assert lines[1].startswith("Arduino AVR Boards")
     assert lines[4].startswith("USBHost")
+
+
+def test_outdated_using_library_with_invalid_version(run_command, data_dir):
+    assert run_command("update")
+
+    # Install latest version of a library library
+    assert run_command("lib install WiFi101")
+
+    # Verifies library is correctly returned
+    res = run_command("outdated")
+    assert res.ok
+    assert "WiFi101" not in res.stdout
+
+    # Changes the version of the currently installed library so that it's
+    # invalid
+    lib_path = Path(data_dir, "libraries", "WiFi101")
+    Path(lib_path, "library.properties").write_text("version=1.0001")
+
+    # Verifies library is correctly returned
+    res = run_command("outdated")
+    assert res.ok
+    lines = [l.strip().split() for l in res.stdout.splitlines()]
+    assert "WiFi101" == lines[1][0]

--- a/test/test_update.py
+++ b/test/test_update.py
@@ -13,6 +13,8 @@
 # software without disclosing the source code of your own applications. To purchase
 # a commercial license, send an email to license@arduino.cc.
 
+from pathlib import Path
+
 
 def test_update(run_command):
     res = run_command("update")
@@ -73,3 +75,25 @@ def test_update_with_url_internal_server_error(run_command, httpserver):
     assert res.failed
     lines = [l.strip() for l in res.stderr.splitlines()]
     assert f"Error updating core and libraries index: downloading index {url}: 500 INTERNAL SERVER ERROR" in lines
+
+
+def test_update_showing_outdated_using_library_with_invalid_version(run_command, data_dir):
+    assert run_command("update")
+
+    # Install latest version of a library
+    assert run_command("lib install WiFi101")
+
+    # Verifies library doesn't get updated
+    res = run_command("update --show-outdated")
+    assert res.ok
+    assert "WiFi101" not in res.stdout
+
+    # Changes the version of the currently installed library so that it's
+    # invalid
+    lib_path = Path(data_dir, "libraries", "WiFi101")
+    Path(lib_path, "library.properties").write_text("version=1.0001")
+
+    # Verifies library gets updated
+    res = run_command("update --show-outdated")
+    assert res.ok
+    assert "WiFi101" in res.stdout

--- a/test/test_upgrade.py
+++ b/test/test_upgrade.py
@@ -13,6 +13,8 @@
 # software without disclosing the source code of your own applications. To purchase
 # a commercial license, send an email to license@arduino.cc.
 
+from pathlib import Path
+
 
 def test_upgrade(run_command):
     # Updates index for cores and libraries
@@ -41,3 +43,25 @@ def test_upgrade(run_command):
     result = run_command("outdated")
     assert result.ok
     assert result.stdout == ""
+
+
+def test_upgrade_using_library_with_invalid_version(run_command, data_dir):
+    assert run_command("update")
+
+    # Install latest version of a library
+    assert run_command("lib install WiFi101")
+
+    # Verifies library is not shown
+    res = run_command("outdated")
+    assert res.ok
+    assert "WiFi101" not in res.stdout
+
+    # Changes the version of the currently installed library so that it's
+    # invalid
+    lib_path = Path(data_dir, "libraries", "WiFi101")
+    Path(lib_path, "library.properties").write_text("version=1.0001")
+
+    # Verifies library gets upgraded
+    res = run_command("upgrade")
+    assert res.ok
+    assert "WiFi101" in res.stdout

